src/IO/S3Common.cpp

```diff
@@ -147,6 +138,12 @@ AuthSettings AuthSettings::loadFromConfig(const std::string & config_elem, const
     };
 }
 
+bool AuthSettings::hasUpdates(const AuthSettings & other) const
+{
+    AuthSettings copy = *this;
+    copy.updateFrom(other);
```

Comment:
why do we need to do this extra copy?

Reply:
Previously `operator ==() const = default` has been used as a function for detection are there any updated. 
But `operator ==() const = default` doesn't fit to it. It brings a bug to the code.

With that context, I had 2 choices:
- write custom `operator ==` -- bad choice, that become totally different fuction
- write custom `bool hasUpdate()` -- good, there are some options how to do it
  - write a big function which carries about each field/member -- hard to read
  - write a small one which reuses `update()` and native `operator ==()` -- I find it easy to read. Yes, there is a copy of a struct.
  - one more alternative it is alter `void update()` to `bool update()`, return value answers to the question were settings been updated.

Reply:
what was the bug if we just use default `==`?

Reply:
It is a good question. In short all is wrong)

1. We compare full set of setting with some diff which we want to apply to the state. It is usually not equal. Exmp: when state has `option<bool>  = false` and the diff has `option<bool>  = null` and vice versa . `operator ==()` return false, but `update()` doesn't change anything. That is real bug apart from the bug in calculation of `static_configuration`.
2. Usually diff has `option<T>` fields and state should not have such fields. We use the same structure for state and diff of the setting. That bring more logic to figure out what the state is and how to apply diff. Example, when in the state `option<bool> no_sign_request = null` that actually means just `bool no_sign_request = false` but it is not clear from the code.

I could have solve p.2 but it would take more efforts. So I just fix p.1.